### ISSUE TITLE
Multiple virt-controller pods on multi-master

### ISSUE THREAD

**Issue by alezzandro** [kubevirt/kubevirt#727]
I want to report that in an envorinment with multi master (2 master in my case) the following command will spawn two virt-controller:

$ kubectl create -f https://github.com/kubevirt/kubevirt/releases/download/v0.2.0/kubevirt.yaml -f manifests/demo-pv.yaml
clusterrole "kubevirt-infra" created
serviceaccount "kubevirt-infra" created
serviceaccount "kubevirt-admin" created
clusterrolebinding "kubevirt-infra" created
clusterrolebinding "kubevirt-infra-cluster-admin" created
customresourcedefinition "virtualmachines.kubevirt.io" created
customresourcedefinition "migrations.kubevirt.io" created
customresourcedefinition "virtualmachinereplicasets.kubevirt.io" created
deployment "virt-controller" created
daemonset "virt-handler" created
daemonset "libvirt" created
service "iscsi-demo-target" created
deployment "iscsi-demo-target-tgtd" created
persistentvolumeclaim "disk-alpine" created
persistentvolume "iscsi-disk-alpine" created

As shown by the following command only one virt-controller pod is marked as "READY":
$ kubectl get --namespace kube-system pods | grep virt
libvirt-9svm2                           2/2       Running   0          6m
libvirt-mb2jl                           2/2       Running   0          6m
libvirt-tfnrp                           2/2       Running   0          6m
virt-controller-b7b8fd9b7-kfjvn         1/1       Running   0          6m
virt-controller-b7b8fd9b7-xl262         0/1       Running   0          6m
virt-handler-gwjjg                      1/1       Running   0          6m
virt-handler-jkxqp                      1/1       Running   0          6m
virt-handler-rs25j                      1/1       Running   0          6m

The other is running but not ready:
$ kubectl logs --namespace kube-system virt-controller-b7b8fd9b7-xl262
level=info timestamp=2018-02-12T13:48:05.392776Z pos=application.go:136 component=virt-controller service=http action=listening interface=0.0.0.0 port=8182

$ kubectl describe --namespace kube-system pod virt-controller-b7b8fd9b7-xl262
Name:           virt-controller-b7b8fd9b7-xl262
Namespace:      kube-system
Node:           k8s-01/172.172.1.101
Start Time:     Mon, 12 Feb 2018 14:42:19 +0100
Labels:         kubevirt.io=virt-controller
                pod-template-hash=636498563
Annotations:    <none>
Status:         Running
IP:             10.233.64.8
Controlled By:  ReplicaSet/virt-controller-b7b8fd9b7
Containers:
  virt-controller:
    Container ID:  docker://500fa5e278feb71cb312d6ff5b5a89457be4927d068e7e2f0c67e70ab6d19446
    Image:         kubevirt/virt-controller:v0.2.0
    Image ID:      docker-pullable://kubevirt/virt-controller@sha256:4f4ce681f5888c21c757a52c9b2458000b8aca89128bfbf10632da7a90630e7b
    Port:          8182/TCP
    Command:
      /virt-controller
      --launcher-image
      kubevirt/virt-launcher:v0.2.0
      --migrator-image
      kubevirt/virt-migrator:v0.2.0
      --port
      8182
    State:          Running
      Started:      Mon, 12 Feb 2018 14:48:05 +0100
    Ready:          False
    Restart Count:  0
    Liveness:       http-get http://:8182/healthz delay=15s timeout=10s period=10s #success=1 #failure=8
    Readiness:      http-get http://:8182/leader delay=15s timeout=10s period=10s #success=1 #failure=3
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kubevirt-infra-token-rn9j9 (ro)
Conditions:
  Type           Status
  Initialized    True 
  Ready          False 
  PodScheduled   True 
Volumes:
  kubevirt-infra-token-rn9j9:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kubevirt-infra-token-rn9j9
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     <none>
Events:
  Type     Reason                 Age               From               Message
  ----     ------                 ----              ----               -------
  Normal   Scheduled              8m                default-scheduler  Successfully assigned virt-controller-b7b8fd9b7-xl262 to k8s-01
  Normal   SuccessfulMountVolume  8m                kubelet, k8s-01    MountVolume.SetUp succeeded for volume "kubevirt-infra-token-rn9j9"
  Normal   Pulling                8m                kubelet, k8s-01    pulling image "kubevirt/virt-controller:v0.2.0"
  Normal   Pulled                 2m                kubelet, k8s-01    Successfully pulled image "kubevirt/virt-controller:v0.2.0"
  Normal   Created                2m                kubelet, k8s-01    Created container
  Normal   Started                2m                kubelet, k8s-01    Started container
  Warning  Unhealthy              2s (x15 over 2m)  kubelet, k8s-01    Readiness probe failed: HTTP probe failed with statuscode: 503

**Comment by cynepco3hahue** [comment]
@alezzandro It was done by design, you can not have two controllers up in the system, otherwise you can get inconsistent results, for example when you will create a new VM two controllers will try to do it, this case you can have a race condition between two controllers.
The system will put the second controller to the ready state, one the leader controller is crashed.

**Comment by rmohr** [comment]
> The system will put the second controller to the ready state, one the leader controller is crashed.

Yes I thought it is a good idea back then, but maybe we are stretching the meaning of "ready" too far here and it is confusing.

**Comment by alezzandro** [comment]
I agree @rmohr, I think you should bring a new up once the existing fails.
Why not using a ReplicaSet?

**Comment by cynepco3hahue** [comment]
@alezzandro We already did discussion you can see @rmohr answer under https://github.com/kubevirt/kubevirt/issues/412#issuecomment-327241295

**Comment by alezzandro** [comment]
Thanks for pointing me to the thread @cynepco3hahue 
BTW any plan for making the virt-controller multi-master so?

**Comment by cynepco3hahue** [comment]
@alezzandro I think that we can achieve it with anti-affinity see [link](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#inter-pod-affinity-and-anti-affinity-beta-feature), but we will need to recognize that we are running multi-master environment first, that can be problematic.

**Comment by alezzandro** [comment]
Excuse me for the silly question: why should you care about number of masters?
Users may want to choose to run multiple active virt-controller for getting high-availability at that level also if they have fewer masters.

Anyway instead of 'raw' affinity/antiaffinity I would look at DaemonSet:
https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/

And in case you want to match the role I saw (never tested btw) that there is a particular affinity rule called "toleration" that match node roles.
Example (from previous link):
 tolerations:
      - key: node-role.kubernetes.io/master

**Comment by cynepco3hahue** [comment]
@alezzandro We do not want to run `virt-controller` instance on each node, we just want to have two instances, in the case if one instance crashed the second will take control. We can add some note about toleration for `virt-controller` pod under user guide.

**Comment by rmohr** [comment]
@alezzandro, @cynepco3hahue  I think you are both right. If there are dedicated master-machines, and we solely want to deploy on them, then the DaemonSet with tolerances and taints is a good solution. If you want to have something like HA withouth the knowledge on how the cluster is set up, one can choose a Deployment, specify an instance count and add an anti-affinity to distribute the controller across the cluster.

**Comment by cynepco3hahue** [comment]
@rmohr We currently run two replicas in our single node environment, in the case when you will have anti-affinity, a scheduler will fail to start one of the pods.

**Comment by fabiand** [comment]
@rmohr can you confirm that this is fixed?

**Comment by rmohr** [comment]
@fabiand there is no bug. It would just make sense to also mark standby controllers as ready, to not confuse everyone what our "ready" status means.

**Comment by rmohr** [comment]
> Users may want to choose to run multiple active virt-controller for getting high-availability at that level also if they have fewer masters.

To answer this question: we follow the same principle like the controller-manager in k8s. There will always ever only be one active controller. We have an election mechanism in the code.
